include/pybind11/detail/value_and_holder.h

```diff
@@ -33,7 +33,11 @@ struct value_and_holder {
 
     template <typename V = void>
     V *&value_ptr() const {
-        return reinterpret_cast<V *&>(vh[0]);
+        if constexpr (std::is_const_v<V>) {
+            return reinterpret_cast<V *&>(const_cast<const void *&>(vh[0]));
```

Comment:
Could we do this unconditionally (without the `std::is_const_v<V>` test)?

To keep the code simpler.